conda/models/records.py

```diff
@@ -322,9 +322,9 @@ def is_unmanageable(self):
     def combined_depends(self):
         from .match_spec import MatchSpec
         result = {ms.name: ms for ms in MatchSpec.merge(self.depends)}
-        result.update({ms.name: ms for ms in MatchSpec.merge(
-            MatchSpec(spec, optional=True) for spec in self.constrains or ()
```

Comment:
This was the main bug that this PR addresses.  That `optional=True` is the problem.